feat(syllabus): add stretch exercises 26, 27, 28 (#517)
CodeYourFuture/cyf-hotel-react: README.md

**Instructions:** Add validation to some fields from exercise 24: the first name and last name must not be empty, the email must contain exactly 1 `@` symbol, and at least one `.` symbol after the `@`; the room ID must be a number between 0 and 100. If the fields do not contain correct information when the 'Submit' button is pressed, display a red error message at the top of the page, but do not clear the text already in the field.

**Test:** An invalid input displays an error message after the 'Submit' button is pressed (e.g. an email like `react@com` is invalid). A valid input shows the correct values at the bottom of the page.

#### 27. Faster validation

**Instructions:** Improve the validation from the last exercise to check for correct values after each input changes, without the 'Submit' button being pressed. Show an 'X' with an explanation next to the field if it has invalid input (e.g. `X first name must not be empty`), and show 'OK' after it has valid input.

**Hint:** Try using the `onChange` function on the `<input>` element

**Test:** The 'Submit' button is initially not clickable, and becomes clickable once every field has the correct input.

#### 28. Date picker

> **Instructions:** Add the [js-datepicker](https://www.npmjs.com/package/js-datepicker) package to your project using `npm install`, and import it at the top of the file. Add different IDs to your 'check in date' and 'check out date' `<input>` elements, then create two date pickers using `const checkInPicker = datepicker(YOUR_ID)` (where `YOUR_ID` is the ID you assigned to your check in/check out date elements). 

Author: great suggestion @SallyMcGrath! added some reflections in 633b315. let me know what you think, and feel free to make further suggestions/update the branch if the wording could be clearer.

**Test:** An invalid input displays an error message after the 'Submit' button is pressed (e.g. an email like `react@com` is invalid). A valid input shows the correct values at the bottom of the page.

> **Reflection:** Validating user input is an important part of any application - without it, you might see unexpected errors when working with the data later. What do you think would happen if you were asked to remove a booking for room number '81', but the room number was not validated and was input as 'eightyOne' or 'EIGHTY ONE'?

Reviewer: Love it.

> **Reflection:** Validating user input is an important part of any application - without it, you might see unexpected errors when working with the data later. What do you think would happen if you were asked to remove a booking for room number '81', but the room number was not validated and was input as 'eightyOne' or 'EIGHTY ONE'?

Reviewer: ```suggestion
**Reflection:** Validating user input is an important part of any application Without checking the input, you might see unexpected errors when working with the data later. 

What do you think would happen if you were asked to remove a booking for room number '81', but the user had typed 'eightyOne' or 'EIGHTY ONE'?
```

**Test:** The 'Submit' button is initially not clickable, and becomes clickable once every field has the correct input.

> **Reflection:** Checking for a valid value after each input changes, instead of once at the end of the page, is a much nicer user experience. If you were a real user of this booking system, would you prefer to fill out all of the fields and find out you made a mistake at the end (like in exercise 26), or would you prefer to be find out you made a mistake after each input?

Reviewer: ```suggestion
**Reflection:** You have used native form validations in HTML. How have you improved this feature with React? 

As a user of this booking system, would you prefer:
-  to find out you made a mistake when you submit the whole form?
-  to find out you made a mistake after each input?
```

#### 28. Date picker

**Instructions:** Add the [js-datepicker](https://www.npmjs.com/package/js-datepicker) package to your project using `npm install`, and import it at the top of the file. Add different IDs to your 'check in date' and 'check out date' `<input>` elements, then create two date pickers using `const checkInPicker = datepicker(YOUR_ID)` (where `YOUR_ID` is the ID you assigned to your check in/check out date elements). 

**Hint:** Read the [js-datepicker usage guide](https://www.npmjs.com/package/js-datepicker#basic-usage)

**Test:** The date picker appears when you click on the 'check in date' and 'check out date' input elements.

> **Reflection:** Using `js-datepicker` in this exercise is a great way to practice installing and working with packages in JavaScript. While HTML has a [native datepicker](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date), using a package will often give you more functions and properties to work with. Packages prevent wasting time 'reinventing the wheel', as someone has already written the code for you! Think about some of the code you have written in this lesson - are there any packages available that might have helped you to complete the exercises? (hint: search 'javascript package' + your idea on Google! e.g. 'javascript package validation')

Reviewer: I think this is a good start, but I also think this reflection pushes the trainee into thinking this is an awesome choice with no downsides. It's a pragmatic choice and one they will often make, but of course we know that typically components that hijack or recreate native elements are not accessible. (And of course often packages are of poor quality in other ways and cause large problems later on, because nobody ever looked inside them to see how they worked, but that's maybe a consideration for later on.)

The majority of all js datepickers available on npm are not accessible (no idea if this one is as I have not set up this project or tested it and it doesn't mention anything in the readme). I would look for a more measured reflection, with some for and against here.

> **Reflection:** Using `js-datepicker` in this exercise is a great way to practice installing and working with packages in JavaScript. While HTML has a [native datepicker](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date), using a package will often give you more functions and properties to work with. Packages prevent wasting time 'reinventing the wheel', as someone has already written the code for you! Think about some of the code you have written in this lesson - are there any packages available that might have helped you to complete the exercises? (hint: search 'javascript package' + your idea on Google! e.g. 'javascript package validation')

Reviewer: I think this is a good start, but I also think this reflection pushes the trainee into thinking this is an awesome choice with no downsides. It's a pragmatic choice and one they will often make, but of course we know that typically components that hijack or recreate native elements are not accessible. (And of course often packages are of poor quality in other ways and cause large problems later on, because nobody ever looked inside them to see how they worked, but that's maybe a consideration for later on.)

The majority of all js datepickers available on npm are not accessible (no idea if this one is as I have not set up this project or tested it and it doesn't mention anything in the readme). I would look for a more measured reflection, with some for and against here.

Author: very good points @SallyMcGrath. I've expanded this reflection in 772bf51 to allow trainees to weigh up the pros/cons of using packages, but I think it may be too long now and tangential to the original focus of the lesson. let me know what you think, happy to simplify or remove it
